### PR TITLE
referral program

### PR DESCRIPTION
**Pull request by a17**

### CODE REVIEW

File: contracts/base/vote/Ve.sol [hunk 1]
```diff
@@ -660,6 +664,10 @@ contract Ve is IERC721, IERC721Metadata, IVe, Reentrancy {
     ++tokenId;
     uint _tokenId = tokenId;
     _mint(_to, _tokenId);
+    if (_ref != 0) {
+      require(_balanceOfNFT(_ref, block.timestamp) >= totalSupply() / 100, 'Need 1% referrer ve balance');
```
Comment (belbix) [inline contracts/base/vote/Ve.sol]: balanceOfNFT() - make it public too

File: contracts/base/vote/Ve.sol [hunk 2]
```diff
@@ -77,6 +77,9 @@ contract Ve is IERC721, IERC721Metadata, IVe, Reentrancy {
   /// @dev ERC165 interface ID of ERC721Metadata
   bytes4 internal constant ERC721_METADATA_INTERFACE_ID = 0x5b5e139f;
 
+  /// @dev Mapping from NFT ID to referrer NFT ID
+  mapping(uint => uint) internal ref;
```
Comment (belbix) [inline contracts/base/vote/Ve.sol]: public

File: contracts/base/vote/Ve.sol [hunk 3]
```diff
@@ -945,4 +955,12 @@ contract Ve is IERC721, IERC721Metadata, IVe, Reentrancy {
   function pointHistory(uint _loc) external view override returns (Point memory) {
     return _pointHistory[_loc];
   }
+
+  function refId(uint _tokenId) external view returns (uint) {
+    return ref[_tokenId];
+  }
+
+  function nftSupply() public view returns (uint) {
+    return tokenId;
```
Comment (belbix) [inline contracts/base/vote/Ve.sol]: make tokenId public and remove it

File: contracts/base/vote/Ve.sol [hunk 4]
```diff
@@ -945,4 +955,12 @@ contract Ve is IERC721, IERC721Metadata, IVe, Reentrancy {
   function pointHistory(uint _loc) external view override returns (Point memory) {
     return _pointHistory[_loc];
   }
+
+  function refId(uint _tokenId) external view returns (uint) {
+    return ref[_tokenId];
```
Comment (belbix) [inline contracts/base/vote/Ve.sol:962]: make ref public and remove it

File: contracts/base/reward/MultiRewardsPoolBase.sol [hunk 5]
```diff
@@ -199,6 +201,13 @@ abstract contract MultiRewardsPoolBase is Reentrancy, IMultiRewardsPool {
       lastEarn[tokens[i]][account] = block.timestamp;
       userRewardPerTokenStored[tokens[i]][account] = rewardPerTokenStored[tokens[i]];
       if (_reward > 0) {
+        /// @dev Extract ref reward 3%
+        if (refId != 0) {
+          uint _refReward = _reward * 3 / 100;
+          _reward -= _refReward;
+          refRewardsFromClaimedPerToken[tokens[i]][refId] += _refReward;
```
Comment (belbix) [inline contracts/base/reward/MultiRewardsPoolBase.sol]: just send it to ref owner

File: contracts/base/reward/Gauge.sol [hunk 6]
```diff
@@ -151,4 +161,36 @@ contract Gauge is IGauge, MultiRewardsPoolBase {
...
+        __earned += _earned(token, IERC721(ve).ownerOf(i)) * 3 / 100;
+      }
+    }
+
+    __earned -= refRewardsReceivedPerToken[token][tokenId];
+
+    return __earned;
+  }
+
+  function getRefReward(uint tokenId, address[] memory tokens) external {
```
Comment (belbix) [inline contracts/base/reward/Gauge.sol]: remove

File: contracts/base/reward/Gauge.sol [hunk 7]
```diff
@@ -151,4 +161,36 @@ contract Gauge is IGauge, MultiRewardsPoolBase {
...
 
+  function earned(address token, address account) external view override returns (uint) {
+    if (IVe(ve).refId(tokenIds[account]) != 0) {
+      return _earned(token, account) * 97 / 100;
+    }
+
+    return _earned(token, account);
+  }
+
+  function refEarned(address token, uint tokenId) public view returns (uint) {
```
Comment (belbix) [inline contracts/base/reward/Gauge.sol]: remove

File: contracts/base/reward/Gauge.sol [hunk 8]
```diff
@@ -142,6 +143,15 @@ contract Gauge is IGauge, MultiRewardsPoolBase {
     if (account == IERC721(ve).ownerOf(_tokenId) && _supply > 0) {
       _adjusted = (totalSupply * IVe(ve).balanceOfNFT(_tokenId) / _supply) * 60 / 100;
     }
+
+    /// @dev Increase boost by 10% for NFT with referrer
+    if (IVe(ve).refId(_tokenId) != 0 && _derived > 0) {
+      _adjusted += _derived * 10 * (_adjusted + _derived) / _derived / 100;
```
Comment (belbix) [inline contracts/base/reward/Gauge.sol]: _adjusted +=  _adjusted / 10

File: contracts/base/reward/MultiRewardsPoolBase.sol [hunk 9]
```diff
@@ -38,6 +38,8 @@ abstract contract MultiRewardsPoolBase is Reentrancy, IMultiRewardsPool {
 
   mapping(address => mapping(address => uint)) public lastEarn;
   mapping(address => mapping(address => uint)) public userRewardPerTokenStored;
+  mapping(address => mapping(uint => uint)) public refRewardsFromClaimedPerToken;
+  mapping(address => mapping(uint => uint)) public refRewardsReceivedPerToken;
```
Comment (belbix) [inline contracts/base/reward/MultiRewardsPoolBase.sol]: move to gauge

File: contracts/base/reward/MultiRewardsPoolBase.sol [hunk 10]
```diff
@@ -190,7 +192,7 @@ abstract contract MultiRewardsPoolBase is Reentrancy, IMultiRewardsPool {
   }
 
   /// @dev Implement restriction checks!
-  function _getReward(address account, address[] memory tokens, address recipient) internal lock virtual {
+  function _getReward(address account, address[] memory tokens, address recipient, uint refId) internal lock virtual {
```
Comment (belbix) [inline contracts/base/reward/MultiRewardsPoolBase.sol]: refAddress

**Review by belbix — CHANGES_REQUESTED** [review]

File: contracts/base/reward/Gauge.sol (hunk 8 above)
Comment (belbix) [reply contracts/base/reward/Gauge.sol]: _derived += _derived / 10

File: contracts/base/reward/MultiRewardsPoolBase.sol (hunk 9 above)
Comment (a17) [reply contracts/base/reward/MultiRewardsPoolBase.sol]: removed

File: contracts/base/reward/Gauge.sol (hunk 8 above)
Comment (a17) [reply contracts/base/reward/Gauge.sol]: _adjusted += _derived / 10;

**Pull request merged**